File: packages/platform/src/ipc/filter.ts
/**
 * 消息过滤规则 IPC 封装 — 过滤规则 CRUD + 启用切换。
 *
 * 后端：壳层 `commands/filter.rs`（InMemoryFilterStore + app::auto_reply::FilterService）。
 *
 * @author agent
 * @created 2026-08-13
 */

import { call } from "./invoke";

/** 过滤类型（与 Rust `FilterType` 对齐）。 */
export type FilterType = "skip_reply" | "skip_notify";

/** 过滤规则（与 Rust `FilterRule` 对齐）。 */
export interface FilterRule {
  id: number;
  account_id: string;
  owner_id: number;
  filter_type: FilterType;
  keyword: string;
  enabled: boolean;
}

/** 过滤类型中文文案。 */
export const FILTER_TYPE_LABELS: Record<FilterType, string> = {
  skip_reply: "跳过自动回复",
  skip_notify: "跳过消息通知",
};

/** 过滤类型选项（供表单使用）。 */
export const FILTER_TYPE_OPTIONS: { value: FilterType; label: string }[] = [
  { value: "skip_reply", label: FILTER_TYPE_LABELS.skip_reply },
  { value: "skip_notify", label: FILTER_TYPE_LABELS.skip_notify },
];

/** 查询账号下的过滤规则。 */
export function filterList(ownerId: number, accountId: string): Promise<FilterRule[]> {
  return call<FilterRule[]>("filter_list", {
    request: { owner_id: ownerId, account_id: accountId },
  });
}

/** 新建过滤规则。 */
export function filterCreate(
  ownerId: number,
  accountId: string,
  rule: Pick<FilterRule, "filter_type" | "keyword">,
): Promise<FilterRule> {
  return call<FilterRule>("filter_create", {
    request: {
      owner_id: ownerId,
      account_id: accountId,
      rule: { ...rule, id: 0, account_id: accountId, owner_id: ownerId, enabled: true },
    },
  });
}

/** 更新过滤规则。 */
export function filterUpdate(
  ownerId: number,
  rule: Pick<FilterRule, "id" | "filter_type" | "keyword" | "account_id">,
): Promise<void> {
  return call<void>("filter_update", {
    request: {
      owner_id: ownerId,
      rule: { ...rule, owner_id: ownerId, enabled: true },
    },
  });
}

/** 切换启用状态。 */
export function filterSetEnabled(ownerId: number, ruleId: number, enabled: boolean): Promise<void> {
  return call<void>("filter_set_enabled", {
    request: { owner_id: ownerId, rule_id: ruleId, enabled },
  });
}

/** 删除过滤规则。 */
export function filterDelete(ownerId: number, ruleId: number): Promise<void> {
  return call<void>("filter_delete", {
    request: { owner_id: ownerId, rule_id: ruleId },
  });
}

File: apps/desktop/src-tauri/src/platforms/xianyu/ipc/filter.rs
//! 消息过滤规则 Tauri commands。

use crate::platforms::xianyu::persist::InMemoryFilterStore;
use app::auto_reply::filter::FilterRule;
use app::auto_reply::FilterService;
use common;
use serde::Deserialize;
use std::sync::Arc;
use tauri::State;

/// 过滤规则服务句柄（setup 时注册到 Tauri 状态）。
pub struct FilterHandle {
    pub store: Arc<InMemoryFilterStore>,
}

#[derive(Debug, Deserialize)]
pub struct FilterListRequest {
    pub owner_id: i64,
    pub account_id: String,
}

#[derive(Debug, Deserialize)]
pub struct FilterCreateRequest {
    pub owner_id: i64,
    pub account_id: String,
    pub rule: FilterRule,
}

#[derive(Debug, Deserialize)]
pub struct FilterUpdateRequest {
    pub owner_id: i64,
    pub rule: FilterRule,
}

#[derive(Debug, Deserialize)]
pub struct FilterEnabledRequest {
    pub owner_id: i64,
    pub rule_id: i64,
    pub enabled: bool,
}

#[derive(Debug, Deserialize)]
pub struct FilterDeleteRequest {
    pub owner_id: i64,
    pub rule_id: i64,
}

#[tauri::command]
pub fn filter_list(
    state: State<'_, FilterHandle>,
    request: FilterListRequest,
) -> common::OpenDeskResult<Vec<FilterRule>> {
    let service = FilterService::new(state.store.as_ref());
    service.list(request.owner_id, &request.account_id)
}

#[tauri::command]
pub fn filter_create(
    state: State<'_, FilterHandle>,
    request: FilterCreateRequest,
) -> common::OpenDeskResult<FilterRule> {
    let service = FilterService::new(state.store.as_ref());
    service.create(request.owner_id, &request.account_id, request.rule)
}

#[tauri::command]
pub fn filter_update(
    state: State<'_, FilterHandle>,
    request: FilterUpdateRequest,
) -> common::OpenDeskResult<()> {
    let service = FilterService::new(state.store.as_ref());
    service
        .update(request.owner_id, &request.rule)
        .map_err(common::OpenDeskError::wrap)
}

#[tauri::command]
pub fn filter_set_enabled(
    state: State<'_, FilterHandle>,
    request: FilterEnabledRequest,
) -> common::OpenDeskResult<()> {
    let service = FilterService::new(state.store.as_ref());
    service.set_enabled(request.owner_id, request.rule_id, request.enabled)
}

#[tauri::command]
pub fn filter_delete(
    state: State<'_, FilterHandle>,
    request: FilterDeleteRequest,
) -> common::OpenDeskResult<()> {
    let service = FilterService::new(state.store.as_ref());
    service.delete(request.owner_id, request.rule_id)
}
